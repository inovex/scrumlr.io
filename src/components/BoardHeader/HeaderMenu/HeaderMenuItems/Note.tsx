import store, {useAppSelector} from "store";
import {ApplicationState} from "types/store";
import {ActionFactory} from "store/action";
import "./HeaderMenuItems.scss";
import classNames from "classnames";

export type NoteProps = {};

export const Note = (props: NoteProps) => {
  const state = useAppSelector((applicationState: ApplicationState) => ({
    board: applicationState.board.data!,
  }));

  return (
    <div className="menu__item-button" id="note">
      <li className="header-menu__item">
        <button
          className="menu__item-button"
          onClick={() => {
            store.dispatch(ActionFactory.editBoard({id: state.board!.id, showNotesOfOtherUsers: !state.board!.showNotesOfOtherUsers}));
          }}
        >
          <div className="item-button__toggle-container">
            <div
              className={classNames(
                "item-button__toggle",
                {"item-button__toggle--left": state.board!.showNotesOfOtherUsers},
                {"item-button__toggle--right": !state.board!.showNotesOfOtherUsers}
              )}
            />
          </div>
          <label className="item-button__label">{state.board!.showNotesOfOtherUsers ? "Hide" : "Show"} notes of other users</label>
        </button>
      </li>
    </div>
  );
};
