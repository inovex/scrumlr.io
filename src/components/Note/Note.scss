@import "constants/style";

$note__border-radius: 16px;
$note__text-margin-bottom: 9px;
$note__box-shadow-color: rgba(0, 87, 255, 0.16);
$note__box-shadow-color-dark-mode: rgba(0, 0, 0, 0.2);

$note__indicator-height: 34px;
$note__indicator-width: 3px;

$note__author-name-size: 12px;
$note__author-image-size: 26px;
$note__author-margin-left: 10px;

.note {
  padding: $padding--default $padding--medium;
  position: relative;
  border-radius: $note__border-radius;
  background-color: $color-white;
  box-shadow: 0 6px 9px 0 $note__box-shadow-color;
  transition: transform 0.3s ease-in-out, box-shadow 0.3s ease-in-out;
  list-style-type: none;
  z-index: $note-z-index;
}
@media (prefers-color-scheme: dark) {
  .note {
    background-color: $color-dark-mode-note;
    box-shadow: 0 6px 9px 0 $note__box-shadow-color-dark-mode;
  }
}
.note:hover {
  transform: scale(1.02);
}

.note--own-card::before {
  content: "";
  display: block;
  position: absolute;
  left: 0;
  width: $note__indicator-width;
  height: $note__indicator-height;
  border-radius: 0 $note__indicator-width $note__indicator-width 0;
  background: var(--accent-color);
}

.note__text {
  margin: 0;
  min-height: 3 * $line-height--medium;
  color: $color-black;
  font-size: $text-size;
  letter-spacing: $letter-spacing--medium;
  line-height: $line-height--medium;
  -webkit-line-clamp: 3;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
  display: -webkit-box;
  margin-bottom: $note__text-margin-bottom;
}
@media (prefers-color-scheme: dark) {
  .note__text {
    color: $color-white;
  }
}

.note__author {
  display: flex;
  align-items: center;
  // reset css from figure-element:
  margin: 0;
  padding: 0;
  border: 0;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
}

.note__author-image {
  height: $note__author-image-size;
  width: $note__author-image-size;
  border-radius: 100%;
}

.note__author-name {
  margin-left: $note__author-margin-left;
  color: $color-dark-gray;
  font-size: $note__author-name-size;
  letter-spacing: $letter-spacing--small;
  line-height: $line-height--small;
}
