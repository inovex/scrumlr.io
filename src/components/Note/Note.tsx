import "./Note.scss";
import avatar from "assets/avatar.png";
import {useSelector} from "react-redux";
import {ApplicationState} from "types/store";
import classNames from "classnames";
import Parse from "parse";

export interface NoteProps {
  text: string;
  authorId: string;
}

const Note = ({text, authorId}: NoteProps) => {
  const state = useSelector((applicationState: ApplicationState) => ({
    board: applicationState.board,
    notes: applicationState.notes,
    users: applicationState.users,
  }));

  return (
    <li className={classNames("note", {"note--own-card": Parse.User.current()?.id === authorId})}>
      <p className="note__text">{text}</p>
      <footer className="note__footer">
        <figure className="note__author" aria-roledescription="author">
          <img className="note__author-image" src={avatar} alt="User" />
          <figcaption className="note__author-name">{state.users.all.filter((user) => user.id === authorId)[0]?.displayName}</figcaption>
        </figure>
      </footer>
    </li>
  );
};
export default Note;
