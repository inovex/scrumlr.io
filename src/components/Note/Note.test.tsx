import React from "react";
import {render} from "@testing-library/react";
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";
import Note from "./Note";

const mockStore = configureStore();

const createNote = (text: string, authorId: string) => {
  const initialState = {
    board: {
      data: {
        columns: [{id: "TestID", name: "Testheader", hidden: false}],
      },
    },
    notes: [],
    users: {
      admins: [],
      basic: [],
      all: [],
    },
  };
  const store = mockStore(initialState);
  return (
    <Provider store={store}>
      <Note text={text} authorId={authorId} />
    </Provider>
  );
};

describe("Note", () => {
  beforeEach(() => {
    window.IntersectionObserver = jest.fn(
      () =>
        (({
          observe: jest.fn(),
          disconnect: jest.fn(),
        } as unknown) as IntersectionObserver)
    );
  });

  describe("should render correctly", () => {
    test("note is present", () => {
      const {container} = render(createNote("Test Text", "Test Author"));
      expect(container.firstChild).toHaveClass("note");
    });

    test("note text is present", () => {
      const {container} = render(createNote("Test Text", "Test Author"));
      expect(container.querySelector(".note").firstChild).toHaveClass("note__text");
    });

    test("note text has correct text", () => {
      const {container} = render(createNote("Test Text", "Test Author"));
      expect(container.firstChild).toHaveTextContent("Test Text");
    });

    test("note footer is present", () => {
      const {container} = render(createNote("Test Text", "Test Author"));
      expect(container.querySelector(".note").lastChild).toHaveClass("note__footer");
    });

    test("note author is present", () => {
      const {container} = render(createNote("Test Text", "Test Author"));
      expect(container.querySelector(".note__footer").firstChild).toHaveClass("note__author");
    });

    test("note author image is present", () => {
      const {container} = render(createNote("Test Text", "Test Author"));
      expect(container.querySelector(".note__author").firstChild).toHaveClass("note__author-image");
    });

    test("note author name is present", () => {
      const {container} = render(createNote("Test Text", "Test Author"));
      expect(container.querySelector(".note__author").lastChild).toHaveClass("note__author-name");
    });
  });

  describe("should have correct style", () => {
    test("show note with correct style", () => {
      const {container} = render(createNote("Test Text", "Test Author"));
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
