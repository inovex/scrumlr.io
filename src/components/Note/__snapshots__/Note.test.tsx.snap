// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note should have correct style show note with correct style 1`] = `
<li
  class="note"
>
  <div
    class="note__content"
  >
    <p
      class="note__text"
    >
      Test Text
    </p>
    <svg
      class="note__edit"
    >
      icon-edit.svg
    </svg>
  </div>
  <footer
    class="note__footer"
  >
    <figure
      aria-roledescription="author"
      class="note__author"
    >
      <img
        alt="User"
        class="note__author-image"
        src="avatar.png"
      />
      <figcaption
        class="note__author-name"
      />
    </figure>
  </footer>
</li>
`;
